layers/water-mesh.js

```diff
@@ -200,10 +235,130 @@ export class WaterMesh extends BufferedMesh {
         this.geometryBindings.delete(key);
       }
     }
+    {
+      const physicsObject = this.physicsObjectsMap.get(key);
+
+      if (physicsObject) {
+        this.physics.removeGeometry(physicsObject);
+        this.physicsObjectsMap.delete(key);
+      }
+    }
+    {
+      const chunkKey = getHashKey(chunk.min.x, chunk.min.y);
+      this.chunkPhysicObjcetMap.delete(chunkKey);
+      this.waterHeightMap.delete(chunkKey);
+    }
     {
       const task = this.gpuTasks.get(key);
       task.cancel();
       this.gpuTasks.delete(key);
     }
   }
+
+  checkWaterContact(chunkPhysicObject, player, waterSurfaceHeight) {
+    // use overlapBox to check whether player contact the water
+    this.physics.enableGeometryQueries(chunkPhysicObject);
+    if (player.avatar) {
+      let collisionIds;
+      const height = player.avatar.height;
+      const width = player.avatar.shoulderWidth
+      if (player.position.y > waterSurfaceHeight) {
+        collisionIds = this.physics.overlapBox(width, height, width, player.position, player.quaternion).objectIds;
+      }
+      else {
+        localVector.set(player.position.x, waterSurfaceHeight, player.position.z);
+        collisionIds = this.physics.overlapBox(width, height, width, localVector, player.quaternion).objectIds;
+      } 
+      for (const collisionId of collisionIds) {
+        if (collisionId === chunkPhysicObject.physicsId) {
+          this.physics.disableGeometryQueries(chunkPhysicObject);
+          return true;
+        }
+      }
+    }
+    this.physics.disableGeometryQueries(chunkPhysicObject);
+    return false;
+  }
+  getSwimDamping(player) {
+    if (this.lastSwimmingHand !== player.avatarCharacterSfx.currentSwimmingHand) {
+      this.lastSwimmingHand = player.avatarCharacterSfx.currentSwimmingHand;
+      if (player.avatarCharacterSfx.currentSwimmingHand !== null) {
+        return normalDamping;
+      }
+    }
+    if (this.swimDamping < maxDamping && this.lastSwimmingHand) {
+      return this.swimDamping *= dampingRate;
+    }
+    else {
+      return maxDamping;
+    }
+  }
+  handleSwimAction(contactWater, player, waterSurfaceHeight) {
+    const swimAction = player.getAction('swim');
+    const hasSwim = !!swimAction;
+    if (contactWater) {
+      this.material.color.setHex( 0x0000ff ); // for testing
+      const characterNeckHeight = player.avatar.height * 0.05;
+      const characterHeadHeight = player.avatar.height * 0.2;

```

Comment:
I think that this might break with different avatars, the head height of an avatar is not always `player.avatar.height * 0.2`
In fact I think the head height is calculated here : https://github.com/upstreet-labs/app/blob/c04fc60e4c24f54b9eb21367ae4008a88bea42a5/avatars/util.mjs#L53
You can do similar calculations with model bones to get the neck height

Reply:
[use ratio of avatar height to add swim action](https://github.com/webaverse/heightfield/pull/6/commits/e849c161454f4190765c3800051901049045e21a)